Add ERC: Vanilla Options for ERC-20 Tokens (#53)
ethereum/ERCs: ERCS/erc-7390.md

---
eip: 7390

> title: Vanilla Option Standard

Reviewer: You will have to remove "standard" from your title (as indicated by the bot.) Every ERC standardizes something, so calling an ERC a standard provides no useful information.

> title: Vanilla Option Standard

Reviewer: How about:

```suggestion
title: Vanilla Options for ERC-20 Tokens
```

I think that's a bit more descriptive?

Author: Yes you're right, we where maybe planning to add support for other token but I'd be a more sophisticated codebase with exceptions etc... Not sure if it will be implemented.

> description: A uniform interface streamlining the creation, management, and execution of vanilla options

Reviewer: ```suggestion
description: An interface for creating, managing, and executing simple time-limited call/put (vanilla) options.
```

How's this? I think it gives a lot more information.

Alice for her part has to allow the spending of **10 TokenC** by calling `approve(address spender, uint256 amount)` on the TokenC's contract, with as parameters the contract's address (`spender`) and for `amount`: **10 \* 10^(TokenC's decimals)**. Then, she can execute `buy` on the contract in order to buy all the options.

We're on the 15th of July, and Alice wants to exercise her options because 1 TokenA is traded at only 10 TokenB! To exercise she has to approve the transferring of **8 TokenA** tokens and call `exercise` on the contract. Bob receives 8 TokenA tokens, and Alice 200 TokenB (8 TokenA \* 25 TokenB). She just made a profit of 200 - 8\*10 = 120 TokenB!

#### Retrieve collateral

Let's say Alice never exercised his option because it wasn't profitable enough for her. To retrieve his collateral, Bob would have to wait for the current time to be greater than `exerciseWindowEnd`. In the examples, this characteristic is set to 2 days, so he would be able to get back his collateral from the 16th of July by simply calling `retrieveExpiredTokens`.

## Rationale

> The proposed ERC option standard provides a simple yet powerful interface for options contracts on Ethereum. By standardizing the interface, it becomes easier for developers to build applications and platforms that support options trading, and users can seamlessly interact with different options contracts across multiple dApps.

Reviewer: ```suggestion
```

This paragraph is justifying the whole proposal; it answers the question "why does ERC-7390 need to exist?" This kind of content belongs in the motivation section.

---
eip: 7390
title: Vanilla Options

> description: An Interface for Vanilla Options

Reviewer: Your title and description need to include more information.Your title should at least hint at what is in this proposal, and your description needs to introduce more information than was available in the title alone.

Author: Improved in commit [47da247](https://github.com/ethereum/ERCs/pull/53/commits/47da24757feff3357df6b75b525604e0f04d05c0), do you think it's better now?

author: Ewan Humbert (@Xeway) <xeway@protonmail.com>, Lassi Maksimainen (@mlalma) <lassi.maksimainen@gmail.com>
discussions-to: https://ethereum-magicians.org/t/erc-7390-vanilla-option-standard/15206
status: Draft
type: Standards Track
category: ERC
created: 2022-09-02
requires: 20
---

## Abstract

> This ERC defines a comprehensive set of functions and events facilitating seamless interactions (creation, management, exercising, etc.) for vanilla options. This standard ensures that options contracts conform to a common interface, facilitating the development of robust options trading platforms and enabling interoperability between dApps and protocols.

Reviewer: Your abstract should give a high-level technical overview so that potential readers get an understanding of both what the proposal is *and* how it works.

I'd also define what a "vanilla option" is. My background is more on the software side, I assumed it was some sort of configuration system, while this proposal is more financial in nature.

Also, all ERCs promote interoperability, so no need to mention that in the abstract.

```suggestion
This ERC defines a comprehensive set of functions and events facilitating seamless interactions (creation, management, exercising, etc.) for vanilla options. It facilitates the development of robust options trading platforms.
```

    event Canceled(uint256 indexed id);
    event PremiumUpdated(uint256 indexed id, uint256 amount);

    function create(VanillaOptionData calldata optionData) external returns (uint256);

    function buy(uint256 id, uint256 amount) external;

    function exercise(uint256 id, uint256 amount) external;

    function retrieveExpiredTokens(uint256 id) external;

    function cancel(uint256 id) external;

    function updatePremium(uint256 id, uint256 amount) external;

    function issuance(uint256 id) external view returns (OptionIssuance memory);
}
```

### State Variable Descriptions

At creation time, user must provide filled instance of `VanillaOptionData` structure that contains all the key information for initializing the option issuance.

#### `side`

**Type: `enum`**

Side of the option. Can take the value `Call` or `Put`. `Call` option gives the option buyer right to exercise any acquired option tokens to buy the `underlying` token at given `strike` price using `strikeToken` from option seller. Similarly, `Put` option gives the option buyer right to sell the `underlying` token to the option seller at `strike` price.

#### `underlyingToken`

> **Type: `address` (`IERC20`)**

Reviewer: I'd expand on what you mean by "(`IERC20`)" somewhere. I think you're saying that `underlyingToken` should point to an ERC-20 contract, but you don't explicitly say that.

```

Emitted when the seller of the option issuance with `id` has retrieved the un-exercised collateral.

#### `Canceled`

```solidity
event Canceled(uint256 indexed id);
```

Emitted when the option issuance with given `id` has been cancelled by the seller.

#### `PremiumUpdated`

```solidity
event PremiumUpdated(uint256 indexed id, uint256 amount);
```

Emitted when seller updates the premium to `amount` for option issuance with given `id`. Note that the updated premium is for the total issuance.

### Concrete Examples

#### Call Option

Let's say Bob sells an **call** option that Alice wants to buy.\
He gives the right to Alice to buy **8 LINK** tokens at **25 USDC** each between **14th of July 2023** and **16th of July 2023**.\
For such a contract, he asks Alice to give him **10 DAI** as a premium.\

To create the contract, he will give the following parameters:

- `side`: **Call**
- `underlyingToken`: **0x514910771AF9Ca656af840dff83E8264EcF986CA** *(LINK's address)*
- `amount`: **8000000000000000000** *(8 \* 10^(LINK's decimals))*
- `strikeToken`: **0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48** *(USDC's address)*
- `strike`: **200000000** *(8 * 25 \* 10^(USDC's decimals))*

> - `premiumToken`: **0x6B175474E89094C44Da98b954EedeAC495271d0F** *(DAI's address)*

Reviewer: I'd recommend using fictitious addresses for your examples, though that isn't strictly required.

This ERC is intended to represent **vanilla** options. However, exotic options can be built on top of this ERC.

## Security Considerations

Contract contains `exerciseWindowStart` and `exerciseWindowEnd` data points. These define the determined time range for the buyer to exercise options. When the current time is greater than `exerciseWindowEnd`, the buyer won't be able to exercise and the seller will be able to retrieve any remaining collateral. 

For preventing clear arbitrage cases when option seller considers the issuance to be of European options, we would strongly advice the option seller to use `updatePremium` call to considerably increase the premium price when exercise window opens. This will make sure that the bots won't be able to buy any remaining options and immediately exercise them for quick profit. If the option issuance is considered to be American, such adjustment is of course not needed. 

Once again, we advise writers to frequently check the underlying token price, and take the best decision for them.

> **Improvement idea:** if two users agreed for an option off-chain and they want to create it on-chain, there is a risk that between the creation of the contract and the purchase by the second user, an on-chain user has already bought the contract. An implementation of ERC-7390 interface might want to add the allowed addresses e.g. to `data` variable or via a separate function call to define the allowed addresses that can buy options.

Reviewer: ```suggestion
**Improvement idea:** if two users agreed for an option off-chain and they want to create it on-chain, there is a risk that between the creation of the contract and the purchase by the second user, an on-chain user has already bought the contract. An implementation of this proposal's interface might want to add the allowed addresses e.g. to `data` variable or via a separate function call to define the allowed addresses that can buy options.
```

## Copyright

> Copyright and related rights waived via [CC0](https://eips.ethereum.org/LICENSE).

Reviewer: ```suggestion
Copyright and related rights waived via [CC0](../LICENSE.md).
```
